utils/misc.ml

```diff
@@ -694,3 +694,194 @@ module MakeHooks(M: sig
   let apply_hooks sourcefile intf =
     fold_hooks !hooks sourcefile intf
 end
+
+
+module Magic_number = struct
+  type version = int
+  type kind =
+      | Exec | Cmi | Cmo | Cma | Cmx | Cmxa | Cmxs | Cmt
+      | Ast_impl | Ast_intf
+
+  let all_kinds =
+    [ Exec; Cmi; Cmo; Cma; Cmx; Cmxa; Cmxs; Cmt; Ast_impl; Ast_intf ]
+
+  type raw = string
+  type info = {
+    kind: kind;
+    version: version;
+  }
+
+  let parse_kind : raw -> kind option = function
+    | "Caml1999X" -> Some Exec
+    | "Caml1999I" -> Some Cmi
+    | "Caml1999O" -> Some Cmo
+    | "Caml1999A" -> Some Cma
+    | "Caml1999Y" -> Some Cmx
+    | "Caml1999Z" -> Some Cmxa
+    | "Caml2007D" -> Some Cmxs
+    | "Caml2012T" -> Some Cmt
+    | "Caml1999M" -> Some Ast_impl
+    | "Caml1999N" -> Some Ast_intf
+    | _ -> None
+
+  let raw_kind : kind -> raw = function
+    | Exec -> "Caml1999X"
+    | Cmi -> "Caml1999I"
+    | Cmo -> "Caml1999O"
+    | Cma -> "Caml1999A"
+    | Cmx -> "Caml1999Y"
+    | Cmxa -> "Caml1999Z"
+    | Cmxs -> "Caml2007D"
+    | Cmt -> "Caml2012T"
+    | Ast_impl -> "Caml1999M"
+    | Ast_intf -> "Caml1999N"
```

Comment:
I like the straightforwardness of the present approach, and the duplication is highly local. Besides, I care about `parse_kind`'s performance more than about `raw_kind`, and unfortunately it is easier and safer to define `parse_kind` from `raw_kind` than the other way around.